src/sciline/pipeline.py

```diff
@@ -316,19 +327,20 @@ def __init__(
         for tp, param in (params or {}).items():
             self[tp] = param
 
-    def insert(self, provider: Provider, /) -> None:
+    def insert(self, provider: Union[ToProvider, Provider], /) -> None:
         """
         Add a callable that provides its return value to the pipeline.
 
         Parameters
         ----------
         provider:
-            Callable that provides its return value. Its arguments and return value
-            must be annotated with type hints.
+            Either a callable that provides its return value. Its arguments
+            and return value must be annotated with type hints.
+            Or a ``Provider`` object that has been constructed from such a callable.
         """
-        if (key := get_type_hints(provider).get('return')) is None:
-            raise ValueError(f'Provider {provider} lacks type-hint for return value')
-        self._set_provider(key, provider)
+        if not isinstance(provider, Provider):
+            provider = Provider.from_function(provider)
+        self._set_provider(provider.deduce_key(), provider)
```

Comment:
Why allow creation of the provider if the key cannot be deduced? Should the check be moved into `from_function`?

Reply:
`from_function` can, and probably should do this check. But we need to check again in `deduce_key` because we cannot deduce non-function provider keys.